examples/energy-management-app/esp32/main/DeviceCallbacks.cpp

```diff
@@ -0,0 +1,42 @@
+/*
+ *
+ *    Copyright (c) 2021-2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "DeviceCallbacks.h"
+
+#include <app/util/util.h>
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/ConcreteAttributePath.h>
+#include <lib/support/logging/CHIPLogging.h>
+
+static const char * TAG = "energy-management-app-callbacks";
+
+using namespace chip;
+using namespace chip::Inet;
+using namespace chip::System;
+using namespace chip::app::Clusters;
+
+void AppDeviceCallbacks::PostAttributeChangeCallback(EndpointId endpointId, ClusterId clusterId, AttributeId attributeId,
+                                                     uint8_t type, uint16_t size, uint8_t * value)
+{
+    ESP_LOGI(TAG, "PostAttributeChangeCallback - Cluster ID: '0x%" PRIx32 "', EndPoint ID: '0x%x', Attribute ID: '0x%" PRIx32 "'",
+             clusterId, endpointId, attributeId);
+
+    ESP_LOGI(TAG, "Current free heap: %u\n", static_cast<unsigned int>(heap_caps_get_free_size(MALLOC_CAP_8BIT)));
+}
```

Comment:
Note EnergyEVSE doesn't send PostAttributeCallbacks since it doesn't use the ember stack for _most_ attributes or commands in the EVSE cluster. However it has a new device callback mechanism to notify the client about. This was PR #30999 which  merged yesterday.

It may need further refactoring to allow that notifier to call out of the delegate into the esp32/main area.

Reply:
Ok